### PR TITLE
Add basic debian packaging

### PR DESCRIPTION
**Pull request by lollipopman**
This adds the required files to build a debian package as well as a
small script to setup the changelog before building.

### CODE REVIEW

File: debian/control [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Source: git-duet
+Section: misc
+Priority: optional
+Maintainer: nobody <nobody@example.com>
```
Comment (rafecolton) [inline debian/control:4]: Hm, I don't think this is right

File: debian/copyright [hunk 2]
```diff
@@ -0,0 +1,25 @@
+Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
+Upstream-Name: git-duet
+Source: https://github.com/git-duet/git-duet
+
+Files: *
+Copyright: 2014, Jesse Szwedko <jesse@szwedko.me>
```
Comment (rafecolton) [inline debian/copyright:6]: Nor this

File: scripts/build-deb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+set -o errexit
+set -o nounset
+
+GIT_DUET_VERSION=$(git describe --tags --always --dirty)
+DEB_VERSION="${GIT_DUET_VERSION}"-1
+# Remove any existing changelog and generate a stub changelog so Debian's
+# packaging can determine the version
+rm -f debian/changelog
+DEBFULLNAME=nobody DEBEMAIL=nobody@example.com dch --create --distribution stable --package git-duet -v "${DEB_VERSION}" "new version"
```
Comment (rafecolton) [inline scripts/build-deb:11]: Nor this

File: debian/compat [hunk 4]
```diff
@@ -0,0 +1 @@
+11
```
Comment (rafecolton) [inline debian/compat:1]: What's this?

**Review by rafecolton — CHANGES_REQUESTED** [review]
I'm not really sure this belongs in the repo, since we're not publishing any debian packages...

**Comment by lollipopman** [comment]
> I'm not really sure this belongs in the repo, since we're not publishing any debian packages...

That decision is certainly up to you folks. I find having packaging committed to a repo's projects helps increase its user base as it allows folks to quickly deploy packages in their companies infrastructure. I put this pull request together to do just that for my employer. Other projects such as [pandoc](https://github.com/jgm/pandoc/releases/tag/2.9.2.1) release their packages, which is quite nice since pandoc, like git-duet, is statically built and doesn't require any other dependencies. Would love to here you thoughts on refining this pull request or closing it, thanks!

**Comment by rafecolton** [comment]
Hm, I definitely hear you, and I'm confident we can come up with something that works.

We already have the homebrew-tap repo in the git-duet org. How about another repo for packaging scripts? Here's what it might look like:

* A single folder per packaging method
* Each folder contains a single bash script with the same name
* A single top-level bash script invokes the script inside the package folder when it is provided as the first positional argument. That top-level script should have a clear usage message.
* If the packaging scripts require dependencies or a Linux environment, they could be done inside a Docker container so as not to affect a user's system. (We already use a Docker container for tests, so I think there is precedent for this.)
* This top-level script could also pull down the latest release binaries from GitHub and validate against a checksum. (We do not currently include checksum files with our releases, but this would be simple to do, and I will add that code if we move forward with this solution.)
* The per-folder script should offer command line flags for configurable values and prompt interactively if those values are not given by flags.
* If the packaging method requires other supporting files like `debian/compat` here, those could be inside the sub-folder OR, if they need to be configurable, they could be written dynamically by the script and gitignored.
* The repo should have Travis tests for each packaging method to make sure each script works and a README within each folder explaining any nuances of the packaging method, if any.

This would allow for folks to contribute scripts necessary for other packaging methods without a huge number of changes to the core git-duet repo.

@jszwedko what do you think?

**Comment by jszwedko** [comment]
Apologies for the delay!

I think building packages for `git-duet` is a good idea and should be relatively simple given the relatively straightforward build requirements.

It seems like, if we will be maintaining the packaging configuration, that it would be reasonable to put in this repo. I commonly expect to see a `debian/` folder at the top-level if there is package configuration in a repo (e.g. [thoughtbot/rcm](https://github.com/thoughtbot/rcm)). I do agree that it simplifies things for users wanting to build the package.

We could do a separate repository as well. I'm not sure what advantages it would have given that we will be maintaining both. I even feel like the homebrew config would be better in here, but it seems like [homebrew expects it to be in a separate repo](https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap) (`homebrew-<repo>`).

Possibly tangential, but I think another way we could solve the problem here is to setup [packagecore](https://github.com/BytePackager/packagecore) to build packages as part of a release which I did somewhat recently for [another project](https://github.com/jszwedko/ec2-metadatafs/blob/master/packagecore.yaml) . It handles generating supporting package configuration as part of the build limiting the amount of expertise in disparate packaging systems needed. You can see the result here: https://github.com/jszwedko/ec2-metadatafs/releases. It is similar in philosophy to [`fpm`](https://github.com/jordansissel/fpm) which could be another option.

I feel like, for a project like `git-duet`, which has a fairly simple build process that results in a static binary, that one of these tools would be sufficient rather than maintaining build configuration for each Linux configuration. The disadvantage is that might be less familiar to people for people wanting to build their own package (rather than just installing a released one).

It might be nice to write a man page if we are going to be building packages (which I think we should!).

Curious to get thoughts!

**Comment by rafecolton** [comment]
@jszwedko I like the idea of using a tool like packagecore rather than maintaining configs for each distro. If there were a single file/directory with those configs or scripts, I wouldn't mind having it in this repo. I have used fpm before and found it fairly straightforward.

I would be curious to see a comparison between the two. Couple initial thoughts and observations:

* FPM looks a lot more popular on GitHub, but that doesn't necessarily mean it's better
* I like that packagecore is python3 so anyone can use it on linux out of the box without needing to install ruby
* I also like that packagecore uses config > code, and in your example, it looks super simple!

> The disadvantage is that might be less familiar to people for people wanting to build their own package

This is true, but not hugely concerning to me. We could write a little documentation on how to build the package locally using packagecore/fpm for someone who wants to build their own.

👍 to writing a man page

**Comment by rafecolton** [comment]
Based on the discussion, I am closing this PR in favor of https://github.com/git-duet/git-duet/issues/90

@lollipopman thanks for bringing this up! Feel free to jump in on the #90 if you are so inclined

**Pull request closed without merging**